File: Indexed Search/Hash Based Indexed Search/Adaptive Thermal Hashing (ATH)/AdaptiveThermalHashing.py
class ATH:

    def __init__(self,size=8):
        self.table=[[] for _ in range(size)]
        self.temp=[0]*size

    def h(self,key):
        return key%len(self.table)

    def insert(self,key,val):

        i=self.h(key)

        self.table[i].append((key,val))

        if len(self.table[i])>3:
            self.rehash()

    def rehash(self):

        old=self.table

        self.table=[[] for _ in range(len(old)*2)]

        for bucket in old:
            for k,v in bucket:
                self.table[self.h(k)].append((k,v))

    def search(self,key):

        i=self.h(key)

        self.temp[i]+=1

        for k,v in self.table[i]:
            if k==key:
                return v

        return None


h=ATH()

h.insert(10,100)

print(h.search(10))

File: Indexed Search/Hash Based Indexed Search/Adaptive Thermal Hashing (ATH)/AdaptiveThermalHashing.cpp
#include <iostream>
#include <vector>
#include <list>

using namespace std;

class ATH {

    vector<list<pair<int,int>>> table;
    vector<int> temp;

public:

    ATH(int n=8){
        table.resize(n);
        temp.resize(n,0);
    }

    int hashKey(int x){
        return abs(x)%table.size();
    }

    void insert(int k,int v){

        int i=hashKey(k);

        table[i].push_back({k,v});

        if(table[i].size()>3)
            rehash();
    }

    void rehash(){

        auto old=table;

        table.clear();

        table.resize(old.size()*2);

        temp.resize(old.size()*2);

        for(auto& b:old)
            for(auto& p:b)
                table[hashKey(p.first)].push_back(p);
    }

    int search(int k){

        int i=hashKey(k);

        temp[i]++;

        for(auto& p:table[i])
            if(p.first==k)
                return p.second;

        return -1;
    }
};

int main(){

    ATH h;

    h.insert(5,50);

    cout<<h.search(5);

}
